src/protocols/secure_channel/SessionResumptionStorage.cpp

```diff
@@ -0,0 +1,189 @@
+    return keyAlloc.SessionResumption(resumptionIdBase64);
+}
+
+CHIP_ERROR SessionResumptionStorage::Save(ScopedNodeId node, ConstResumptionIdView resumptionId,
+                                          const Crypto::P256ECDHDerivedSecret & sharedSecret, const CATValues & peerCATs)
+{
+    DefaultStorageKeyAllocator keyAlloc;
+    {
+        // Save session state into key: /f/<fabricIndex>/s/<nodeId>
+        uint8_t buf[MaxStateSize()];
+
+        TLV::TLVWriter writer;
+        writer.Init(buf);
+
+        TLV::TLVType outerType;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outerType));
+
+        ReturnErrorOnFailure(writer.Put(kResumptionIdTag, resumptionId));
+        ReturnErrorOnFailure(writer.Put(kSharedSecretTag, ByteSpan(sharedSecret.ConstBytes(), sharedSecret.Length())));
+        CATValues::Serialized cat;
+        peerCATs.Serialize(cat);
+        ReturnErrorOnFailure(writer.Put(kCATTag, ByteSpan(cat)));
+
+        ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+        const auto len = writer.GetLengthWritten();
+        VerifyOrDie(CanCastTo<uint16_t>(len));
+
+        ReturnErrorOnFailure(mStorage->SyncSetKeyValue(StorageKey(keyAlloc, node), buf, static_cast<uint16_t>(len)));
+    }
+
+    {
+        // Save a link from resumptionId to node, in key: /f/<fabricIndex>/r/<resumptionId>
+        uint8_t buf[MaxLinkSize()];
+
+        TLV::TLVWriter writer;
+        writer.Init(buf);
+
+        TLV::TLVType outerType;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outerType));
+        ReturnErrorOnFailure(writer.Put(kFabricIndexTag, node.GetFabricIndex()));
+        ReturnErrorOnFailure(writer.Put(kPeerNodeIdTag, node.GetNodeId()));
+        ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+        const auto len = writer.GetLengthWritten();
+        VerifyOrDie(CanCastTo<uint16_t>(len));
+
+        ReturnErrorOnFailure(mStorage->SyncSetKeyValue(StorageKey(keyAlloc, resumptionId), buf, static_cast<uint16_t>(len)));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SessionResumptionStorage::FindByScopedNodeId(ScopedNodeId node, ConstResumptionIdView & resumptionId,
+                                                        Crypto::P256ECDHDerivedSecret & sharedSecret, CATValues & peerCATs)
+{
+    DefaultStorageKeyAllocator keyAlloc;
+    uint8_t buf[MaxStateSize()];
+    uint16_t len = static_cast<uint16_t>(MaxStateSize());
+
+    ReturnErrorOnFailure(mStorage->SyncGetKeyValue(StorageKey(keyAlloc, node), buf, len));
+
+    TLV::ContiguousBufferTLVReader reader;
+    reader.Init(buf, len);
+
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+    TLV::TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    ReturnErrorOnFailure(reader.Next(kResumptionIdTag));
+    ReturnErrorOnFailure(reader.Get(resumptionId));
```

Comment:
This is why we can't have the implicit FixedSpan(Span) constructor.  If an incorrect length resumptionId is persisted, this will cause a VerifyOrDie reset loop when we try and fail to construct the resumptionId FixedSpan.